File: server/compose/types/page_layout.go
package types

import (
	"database/sql/driver"
	"encoding/json"
	"strconv"
	"strings"
	"time"

	"github.com/cortezaproject/corteza/server/pkg/locale"
	"github.com/cortezaproject/corteza/server/pkg/sql"

	"github.com/cortezaproject/corteza/server/pkg/filter"
)

type (
	PageLayout struct {
		ID          uint64 `json:"pageLayoutID,string"`
		NamespaceID uint64 `json:"namespaceID,string"`
		PageID      uint64 `json:"pageID,string"`
		ParentID    uint64 `json:"parentID,string"`
		Handle      string `json:"handle"`
		Primary     bool   `json:"primary"`

		Weight int `json:"weight"`

		Meta *PageLayoutMeta `json:"meta"`

		Config PageLayoutConfig `json:"config"`
		Blocks PageBlocks       `json:"blocks"`

		Labels map[string]string `json:"labels,omitempty"`

		OwnedBy uint64 `json:"ownedBy,string"`

		CreatedAt time.Time  `json:"createdAt,omitempty"`
		UpdatedAt *time.Time `json:"updatedAt,omitempty"`
		DeletedAt *time.Time `json:"deletedAt,omitempty"`
	}

	PageLayoutMeta struct {
		// Warning: value of this field is now handled via resource-translation facility
		//          struct field is kept for the convenience for now since it allows us
		//          easy encoding/decoding of the outgoing/incoming values
		Name string `json:"name"`

		// Warning: value of this field is now handled via resource-translation facility
		//          struct field is kept for the convenience for now since it allows us
		//          easy encoding/decoding of the outgoing/incoming values
		Description string `json:"description"`
	}

	PageLayoutButton struct {
		Label   string `json:"label"`
		Enabled bool   `json:"enabled"`
	}

	PageLayoutConfig struct {
		Visibility PageLayoutVisibility `json:"visibility"`
		Actions    []PageLayoutAction   `json:"actions"`
	}

	PageLayoutVisibility struct {
		Expression string   `json:"expression"`
		Roles      []string `json:"roles"`
	}

	PageLayoutAction struct {
		ActionID  uint64 `json:"actionID,string"`
		Placement string
		Meta      any

		// Warning: value of this field is now handled via resource-translation facility
		//          struct field is kept for the convenience for now since it allows us
		//          easy encoding/decoding of the outgoing/incoming values
		Label string `json:"label"`

		// Kind and Params specify the action's behavior and the parameters it
		// can use for execution
		Kind   string
		Params any
	}

	PageLayoutFilter struct {
		NamespaceID uint64 `json:"namespaceID,string"`
		PageID      uint64 `json:"pageID,string,omitempty"`
		Primary     bool   `json:"primary,omitempty"`
		Handle      string `json:"handle"`
		Name        string `json:"name"`
		Query       string `json:"query"`

		LabeledIDs []uint64          `json:"-"`
		Labels     map[string]string `json:"labels,omitempty"`

		Deleted filter.State `json:"deleted"`

		// Check fn is called by store backend for each resource found function can
		// modify the resource and return false if store should not return it
		//
		// Store then loads additional resources to satisfy the paging parameters
		Check func(*PageLayout) (bool, error) `json:"-"`

		// Standard helpers for paging and sorting
		filter.Sorting
		filter.Paging
	}
)

func (m PageLayout) Clone() *PageLayout {
	c := &m
	return c
}

func (p *PageLayout) decodeTranslations(tt locale.ResourceTranslationIndex) {
	var aux *locale.ResourceTranslation

	// @note not doing blocks because they are simply copied from the page's index

	for i, action := range p.Config.Actions {
		actionID := locale.ContentID(action.ActionID, i)
		rpl := strings.NewReplacer(
			"{{actionID}}", strconv.FormatUint(actionID, 10),
		)

		if aux = tt.FindByKey(rpl.Replace(LocaleKeyPagePageBlockBlockIDTitle.Path)); aux != nil {
			p.Config.Actions[i].Label = aux.Msg
		}
	}
}

func (p *PageLayout) encodeTranslations() (out locale.ResourceTranslationSet) {
	out = make(locale.ResourceTranslationSet, 0, 8)

	// @note not doing blocks because they are simply copied from the page's index

	// Actions
	for i, action := range p.Config.Actions {
		actionID := locale.ContentID(action.ActionID, i)
		rpl := strings.NewReplacer(
			"{{actionID}}", strconv.FormatUint(actionID, 10),
		)

		out = append(out, &locale.ResourceTranslation{
			Resource: p.ResourceTranslation(),
			Key:      rpl.Replace(LocaleKeyPageLayoutConfigActionsActionIDLabel.Path),
			Msg:      action.Label,
		})
	}

	return
}

// FindByHandle finds pageLayout by it's handle
func (set PageLayoutSet) FindByHandle(handle string) *PageLayout {
	for i := range set {
		if set[i].Handle == handle {
			return set[i]
		}
	}

	return nil
}

func (bb *PageLayoutConfig) Scan(src any) error          { return sql.ParseJSON(src, bb) }
func (bb PageLayoutConfig) Value() (driver.Value, error) { return json.Marshal(bb) }

func (vv *PageLayoutMeta) Scan(src any) error           { return sql.ParseJSON(src, vv) }
func (vv *PageLayoutMeta) Value() (driver.Value, error) { return json.Marshal(vv) }

File: lib/js/src/compose/types/page-layout.ts
import { merge } from 'lodash'
import { Apply, CortezaID, ISO8601Date, NoID } from '../../cast'
import { PageBlock } from './page-block/base'
import { Button } from './page-block/types'

export type PageLayoutInput = PageLayout | Partial<PageLayout>

interface PageLayoutConfig {
  visibility: Visibility;
  buttons: {
    back: Button;
    delete: Button;
    new: Button;
    clone: Button;
    edit: Button;
    submit: Button;
  };
  actions: Action[];
  // Only used for record pages
  useTitle: Boolean;
}

interface Action {
  kind: string;
  enabled: boolean;
  placement: string;
  params: unknown;
  meta: ActionMeta;
}

interface ActionMeta {
  label: string;
  style: {
    variant: string;
  }
}

interface Visibility {
  expression: string;
  roles: string[];
}

interface Meta {
  title: string;
  description: string;
}

export class PageLayout {
  public pageLayoutID = NoID;
  public namespaceID = NoID;
  public pageID = NoID
  public handle = '';

  public weight = 0;

  public blocks: (Partial<PageBlock>)[] = [];

  public config: PageLayoutConfig = {
    visibility: {
      expression: '',
      roles: [],
    },
    buttons: {
      back: { enabled: true },
      delete: { enabled: true },
      clone: { enabled: true },
      new: { enabled: true },
      edit: { enabled: true },
      submit: { enabled: true },
    },
    actions: [],
    useTitle: false,
  }

  public meta: Meta = {
    title: '',
    description: ''
  };

  public createdAt?: Date = undefined;
  public updatedAt?: Date = undefined;
  public deletedAt?: Date = undefined;

  public ownedBy = NoID;

  constructor (pl?: PageLayoutInput) {
    this.apply(pl)
  }

  apply (pl?: PageLayoutInput): void {
    if (!pl) return

    Apply(this, pl, CortezaID, 'pageLayoutID', 'namespaceID', 'pageID', 'ownedBy')
    Apply(this, pl, String, 'handle')
    Apply(this, pl, Number, 'weight')
    Apply(this, pl, ISO8601Date, 'createdAt', 'updatedAt', 'deletedAt')

    this.blocks = (pl.blocks || []).map(({ blockID, xywh, meta }) => ({ blockID, xywh, meta }))

    if (pl.meta) {
      this.meta = { ...this.meta, ...pl.meta }
    }

    if (pl.config) {
      this.config = merge({}, this.config, pl.config)
    }
  }

  clone (): PageLayout {
    return new PageLayout(JSON.parse(JSON.stringify(this)))
  }

  addAction () {
    this.config.actions.push({
      kind: 'toLayout',
      placement: 'end',
      enabled: true,
      params: {
        pageLayoutID: '',
      },
      meta: {
        label: '',
        style: {
          variant: 'primary'
        }
      },
    } as Action)
  }

  /**
   * Returns resource ID
   */
  get resourceID (): string {
    return `${this.resourceType}:${this.pageLayoutID}`
  }

  /**
   * Resource type
   */
  get resourceType (): string {
    return 'compose:page-layout'
  }

  export (): PageLayoutInput {
    return {
      blocks: this.blocks,
      meta: this.meta,
    }
  }
}
